### PR TITLE
Add migration test with rng device

### PR DESCRIPTION
**Pull request by fangge1212**
The following rng backend are covered:
tcp(bind, connect), udp(bind_connect, connect), rdm, builtin

Signed-off-by: Fangge Jin <fjin@redhat.com>

### CODE REVIEW

**Comment by fangge1212** [comment]
@dzhengfy @Yingshun 
Please help me review this pr.

File: libvirt/tests/cfg/migration/migrate_virtual_devices.cfg [hunk 1]
```diff
@@ -0,0 +1,41 @@
+- migrate_virtual_devices:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/migration/migrate_virtual_devices.cfg:1]: generally keyword is separated by '.' rather than '_'

**Comment by chunfuwen** [comment]
suggest rename files to more exactly such as :migrate_rng_device.py or migrate_rng_device.cfg

File: libvirt/tests/src/migration/migrate_virtual_devices.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    :param backend_dev:
+    :param backend_source:
+    :param backend_protocol:
+
+    """
+
+    def __init__(self, test, env, params, *args, **dargs):
+        for k, v in iteritems(dict(*args, **dargs)):
+            params[k] = v
+        params['hotplug'] = params.get("hotplug", "no")
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/migrate_virtual_devices.py:34]: I assume there are not necessary to override values in params

Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_virtual_devices.py:34]: This is just convenient for us to pass a parameter without re-do bootstrap.

**Comment by fangge1212** [comment]
> suggest rename files to more exactly such as :migrate_rng_device.py or migrate_rng_device.cfg

We planned several py files: migration_memory.py, migration_cpu.py, migration_virtual_disk.py, migration_virtual_network.py, migration_virtual_devices.py, migration_misc.py.  migration_virtual_devices.py will include all the left virtual devices except for memory/cpu/disk/network.

File: libvirt/tests/cfg/migration/migrate_virtual_devices.cfg (hunk 1 above)
Comment (fangge1212) [reply libvirt/tests/cfg/migration/migrate_virtual_devices.cfg:1]: Do you mean it should be renamed to: migrate.virtual_devices?

Comment (fangge1212) [reply libvirt/tests/cfg/migration/migrate_virtual_devices.cfg]: Updated the case pattern

Comment (fangge1212) [reply libvirt/tests/cfg/migration/migrate_virtual_devices.cfg]: Simplify it to "migrate"

**Review by chunfuwen — APPROVED** [review]
LGTM

**Comment by dzhengfy** [comment]
> > suggest rename files to more exactly such as :migrate_rng_device.py or migrate_rng_device.cfg
> 
> We planned several py files: migration_memory.py, migration_cpu.py, migration_virtual_disk.py, migration_virtual_network.py, migration_virtual_devices.py, migration_misc.py. migration_virtual_devices.py will include all the left virtual devices except for memory/cpu/disk/network.

@fangge1212 what is migration_misc.py for?

**Review by dzhengfy — COMMENTED** [review]
Please help provide meaningful commit message and fill in the PR description field. Thanks.

**Comment by dzhengfy** [comment]
@fangge1212 Please also help provide test results.

**Comment by fangge1212** [comment]
> > > suggest rename files to more exactly such as :migrate_rng_device.py or migrate_rng_device.cfg
> > 
> > 
> > We planned several py files: migration_memory.py, migration_cpu.py, migration_virtual_disk.py, migration_virtual_network.py, migration_virtual_devices.py, migration_misc.py. migration_virtual_devices.py will include all the left virtual devices except for memory/cpu/disk/network.
> 
> @fangge1212 what is migration_misc.py for?

For all the other tests that can't be covered by above mentioned py files.
Actually I'm not sure whether someone will need this.

**Comment by fangge1212** [comment]
Test result:
UDP rng tests failed: logging in vm serial console times out when check vm uptime before migration back.
I'm debugging it.

Test name | Duration | Status
-- | -- | --
virtual_devices.rng.backend_builtin.native_tls.postcopy | 10 min | Fixed
virtual_devices.rng.backend_builtin.plaintext.precopy | 7 min 57 sec | Fixed
virtual_devices.rng.backend_rdm.native_tls.postcopy | 10 min | Fixed
virtual_devices.rng.backend_rdm.plaintext.precopy | 7 min 52 sec | Fixed
virtual_devices.rng.backend_tcp.bind_mode.native_tls.postcopy | 10 min | Passed
virtual_devices.rng.backend_tcp.bind_mode.plaintext.precopy | 8 min 14 sec | Passed
virtual_devices.rng.backend_tcp.connect_mode.native_tls.postcopy | 10 min | Passed
virtual_devices.rng.backend_tcp.connect_mode.plaintext.precopy | 7 min 58 sec | Fixed
virtual_devices.rng.backend_udp.bind_connect_mode.native_tls.postcopy | 20 min | Failed
virtual_devices.rng.backend_udp.bind_connect_mode.plaintext.precopy | 20 min | Failed
virtual_devices.rng.backend_udp.connect_mode.native_tls.postcopy | 20 min | Failed
virtual_devices.rng.backend_udp.connect_mode.plaintext.precopy | 20 min | Failed

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+            logging.debug("rng_attr=%s, set_value=%s, get_value=%s", rng_attr, set_value, get_value)
+            if set_value and set_value != get_value:
+                if rng_attr == 'backend':
+                    for bak_attr in xml_set.backend.__slots__:
+                        set_backend, get_backend = get_compare_values(xml_set.backend, xml_get.backend, bak_attr)
+                        if set_backend and set_backend != get_backend:
+                            if bak_attr == 'source':
+                                set_source = xml_set.backend.source
+                                get_source = xml_get.backend.source
+                                find = False
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/migrate_with_virtual_devices.py:279]: for and if statements construct more depth logic, and is too hard to understand and maintain.
I wonder if by reusing some common lib or itself can be simplified

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            cmd = "cat /dev/urandom | nc -k -l %s 1024" % server
+        elif mode == 'client':
+            cmd = "cat /dev/urandom | nc %s 1024" % server
+        else:
+            self.test.error("Uknown rng source mode")
+            return
+        rngjob = utils_misc.AsyncJob(cmd, kill_func=self.func_kill_rng_source)
+        self.bgjobs.append(rngjob)
+
+        # TODO: Need to check the tcp connection is established|listening
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/migrate_with_virtual_devices.py:195]: remove commented out code

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        """
+        Create an external source for tcp backend rng device.
+        Also open port of rng server in firewalld
+
+        :param server: ip of the rng tcp server
+        :param mode: the mode of the external rng source: 'client' or 'server'
+                     Note:
+                     Set it to 'client' if vm rng backend_type is 'bind';
+                     Set it to 'server' if vm rng backend_type is 'connect'
+        :param cleanup: True to cleanup rng src instead of creating
```
Comment (Yingshun) [inline libvirt/tests/src/migration/migrate_with_virtual_devices.py:169]: do you need to remove 'cleanup' ?

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    def _create_tcp_rng_source(self, server, mode=None):
+        """
+        Create an external source for tcp backend rng device.
+        Also open port of rng server in firewalld
+
+        :param server: ip of the rng tcp server
+        :param mode: the mode of the external rng source: 'client' or 'server'
+                     Note:
+                     Set it to 'client' if vm rng backend_type is 'bind';
+                     Set it to 'server' if vm rng backend_type is 'connect'
```
Comment (Yingshun) [inline libvirt/tests/src/migration/migrate_with_virtual_devices.py:168]: I wonder if we can use backend_type directly. It looks a little complex for now...

Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:168]: If vm rng backend_type is 'bind', then vm is server, the rng source created here is client;
if vm rng backend_type is 'connect', then vm is client, the rng source created here is server.

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 3 above)
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:273]: This part is copied from libvirt_rng.py, should we @ the original author here?

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 4 above)
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: Done

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            if self.params['backend_type'] == 'tcp':
+                for source in source_list:
+                    if not source.get('host') and source['mode'] == 'connect':
+                        source.update(host=self.params['local_ip'])
+                    if not source.get('host') and source['mode'] == 'bind':
+                        source.update(host='0.0.0.0')
+            backend.source = source_list
+        if self.params['backend_protocol']:
+            backend.backend_protocol = self.params['backend_protocol']
+        self.rng_xml.backend = backend
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:72]: The part is copied from libvirt_rng.py, should we @ the original author here?

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+        # check if xml attr same with checking
+        try:
+            rng_index = xml_devices.index(rng_devices[0])
+            xml_get = xml_devices[rng_index]
+
+            if not rng_present:
+                return False
+        except IndexError:
+            if rng_present:
+                logging.error("Rng xml is not in dumpxml, this is not expected")
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: "rng_present" is the expected result, "IndexError" indicts it is not present actually, so the error is correct.

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+
+        # Create rng source - client
+        if self.params['rng_source_mode'] == 'bind':
+            self._create_tcp_rng_source(self.params['migrate_source_host'], mode='client')
+
+        # Check rng device
+        if not self.check_rng_in_dumpxml(self.main_vm,
+                                         self.rng_xml,
+                                         rng_present=True):
+            self.test.fail("Failed when check rng in dumpxml after hotplug")
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: I'll try

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        rng_devices = xml_devices.by_device_tag("rng")
+        logging.debug("rng_devices is %s", rng_devices)
+
+        # check if xml attr same with checking
+        try:
+            rng_index = xml_devices.index(rng_devices[0])
+            xml_get = xml_devices[rng_index]
+
+            if not rng_present:
+                return False
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:230]: I'll try

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: Done

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: Done

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+    :param backend_protocol:
+
+    """
+
+    def __init__(self, test, env, params, *args, **dargs):
+        for k, v in iteritems(dict(*args, **dargs)):
+            params[k] = v
+        params['hotplug'] = params.get("hotplug", "no")
+        params['rng_source_mode'] = params.get("rng_source_mode")
+        params['backend_type'] = params.get("backend_type")
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:128]: Set the parameter to none if it is not set in params

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+    Do migration with rng device
+
+    :param hotplug: hotplug device or not
+    :param rng_source_mode:
+    :param backend_type:
+    :param rng_model:
+    :param backend_model:
+    :param backend_dev:
+    :param backend_source:
+    :param backend_protocol:
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:119]: done

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+            return (set_value, get_value)
+
+        match = True
+        for rng_attr in xml_set.__slots__:
+            set_value, get_value = get_compare_values(xml_set, xml_get, rng_attr)
+            logging.debug("rng_attr=%s, set_value=%s, get_value=%s", rng_attr, set_value, get_value)
+            if set_value and set_value != get_value:
+                if rng_attr == 'backend':
+                    for bak_attr in xml_set.backend.__slots__:
+                        set_backend, get_backend = get_compare_values(xml_set.backend, xml_get.backend, bak_attr)
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:395]: I'll do it, please review other parts first.

[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: Done

**Comment by fangge1212** [comment]
@Yingshun  I have updated the pr according to the comments, pls help me review again.

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 6 above)
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py:272]: I added a new method and changed it to:
         # Need to create rng server before vm starts if
         # vm rng mode is 'client'
         if self._vm_rng_mode() == 'client':
             self._create_tcp_rng_source(self.migrate_source_host,
                                         mode='server')

**Comment by fangge1212** [comment]
Test result:

Test name | Duration | Status
-- | -- | --
virtual_devices.rng.backend_builtin.native_tls.postcopy | 11 min | Passed
virtual_devices.rng.backend_builtin.plaintext.precopy | 8 min 39 sec | Passed
virtual_devices.rng.backend_rdm.native_tls.postcopy | 11 min | Passed
virtual_devices.rng.backend_rdm.plaintext.precopy | 8 min 35 sec | Passed
virtual_devices.rng.backend_tcp.bind_mode.native_tls.postcopy | 12 min | Passed
virtual_devices.rng.backend_tcp.connect_mode.plaintext.precopy | 9 min 8 sec | Passed
virtual_devices.rng.backend_udp.bind_connect_mode.plaintext.precopy | 8 min 54 sec | Passed
virtual_devices.rng.backend_udp.connect_mode.native_tls.postcopy | 11 min | Passed

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: I updated this

**Review by chunfuwen — COMMENTED** [review]
I assume we have  one guideline for commit message, please update commit message accordingly

**Comment by fangge1212** [comment]
> I assume we have one guideline for commit message, please update commit message accordingly

I'm not sure what specific points you mean, but I just updated the commit message. Please help to review again.

**Comment by dzhengfy** [comment]
@fangge1212 , please help add comment with 'xxx depends on PR <xxx>' explicitly which will help the maintainer easily detect the dependency. Thanks a lot.

**Comment by dzhengfy** [comment]
This PR depends on https://github.com/avocado-framework/avocado-vt/pull/2709

**Review by chunfuwen — APPROVED** [review]
LGTM

**Comment by Yingshun** [comment]
@fangge1212 May I ask the test results for the modular daemon env? We have almost finished the debugging the existing tests.

**Comment by fangge1212** [comment]
> @fangge1212 May I ask the test results for the modular daemon env? We have almost finished the debugging the existing tests.

Hi @Yingshun 
Thanks for your reminder. The remote access protocol is set to "tcp" by default in migration_template.py(it can be changed to ssh or tls in base.cfg or <test>.cfg, I can add the ssh logic for modular daemon later), and there is no need to check libvirtd.log for this pr. Is there other places that need to be updated for modular daemon?

**Comment by fangge1212** [comment]
Set migrate_desturi_proto to ssh, and run it modular daemon mode:
# avocado run --vt-type libvirt --vt-machine-type q35 migrate.virtual_devices.rng.backend_rdm.plaintext.precopy --vt-connect-uri  qemu:///system
JOB ID     : f95c37f185283b284ffc767fc772a8ee5254cfac
JOB LOG    : /root/avocado/job-results/job-2020-09-30T23.03-f95c37f/job.log
 (1/1) type_specific.io-github-autotest-libvirt.migrate.virtual_devices.rng.backend_rdm.plaintext.precopy: PASS (441.93 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 443.78 s

**Comment by fangge1212** [comment]
> @fangge1212 May I ask the test results for the modular daemon env? We have almost finished the debugging the existing tests.

@Yingshun I just added test result for modular daemon

**Review by Yingshun — APPROVED** [review]

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+
+    :param hotplug: hotplug device or not
+    :param rng_source_mode: vm rng device source mode
+    :param backend_type: vm rng device backend type
+    :param rng_model: vm rng device model
+    :param backend_model: vm rng device backend model
+    :param backend_dev: vm rng device backend dev
+    :param backend_source: vm rng device backend source
+    :param backend_protocol: vm rng device backend protocol
+
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_with_virtual_devices.py:166]: Remove this blank line, please.

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py (hunk 14 above)
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: done

File: libvirt/tests/src/migration/migrate_with_virtual_devices.py [hunk 15]
```diff
@@ -0,0 +1,585 @@
...
+    Raise when device attributes in domain dumpxml is not same as set values
+    """
+
+    def __str__(self):
+        super_str = super(DeviceXMLMismatchError, self).__str__()
+        return (super_str +
+                "Device domain xml is not same as set")
+
+
+class migration_with_rng(MigrationTemplate):
```
[the remark this answers is not in the document]
Comment (fangge1212) [reply libvirt/tests/src/migration/migrate_with_virtual_devices.py]: done. Changed to MigrationWithRng

**Review by dzhengfy — APPROVED** [review]

**Pull request merged**